src/java_bytecode/java_string_literals.cpp

```diff
@@ -0,0 +1,211 @@
+/*******************************************************************\
+
+Module: Java string literal processing
+
+Author: Chris Smowton, chris.smowton@diffblue.com
+
+\*******************************************************************/
+
+#include "java_string_literals.h"
+#include "java_root_class.h"
+#include "java_types.h"
+#include "java_utils.h"
+
+#include <linking/zero_initializer.h>
+
+#include <util/arith_tools.h>
+#include <util/namespace.h>
+#include <util/unicode.h>
+
+#include <iomanip>
+#include <sstream>
+
+/// Replace non-alphanumeric characters with escape sequences. Anything non-
+/// alphanumeric is replaced by _xx where xx is a hex character code, and a
+/// literal underscore is doubled (replaced by __).
+/// \param to_escape: string to escape
+/// \return string with non-alphanumeric characters replaced by escape
+///   sequences.
+static std::string escape_non_alnum(const std::string &to_escape)
+{
+  std::ostringstream escaped;
+  for(auto &ch : to_escape)
+  {
+    if(ch=='_')
+      escaped << "__";
+    else if(isalnum(ch))
+      escaped << ch;
+    else
+      escaped << '_'
+              << std::hex
+              << std::setfill('0')
+              << std::setw(2)
+              << (unsigned int)ch;
+  }
+  return escaped.str();
+}
```

Comment:
Moving those functions into a separate file should be done in a different commit. Right now it looks similar, but not identical.